### PR TITLE
PHOENIX-1661 Implement built-in functions for JSON

### PR DESCRIPTION
**Pull request by ictwanglei**
1. Add some functional  files: JsonArrayElementsFunction.java  JsonArrayLengthFunction.java JsonEachFunction.java  JsonObjectKeysFunction.java  JsonPopulateRecordFunction.java JsonPopulateRecordSetFunction.java. ArrayToJsonFunction.java

2. Add a unit test file: JsonFunctionTest.java

3. Modify file PhoenixJson.java to which I added some assistant functions.

@JamesRTaylor @twdsilva @AakashPradeep please review the changelist.

### CODE REVIEW

File: phoenix-core/src/main/java/org/apache/phoenix/expression/function/ArrayToJsonFunction.java [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+
+import java.io.IOException;
+import java.sql.SQLException;
+import java.util.List;
+
+
+@FunctionParseNode.BuiltInFunction(name=ArrayToJsonFunction.NAME,  args={
+        @FunctionParseNode.Argument(allowedTypes={PVarchar.class})} )
+public class ArrayToJsonFunction extends ScalarFunction {
+    public static final String NAME = "Array_To_Json";
```
Comment (AakashPradeep) [inline phoenix-core/src/main/java/org/apache/phoenix/expression/function/ArrayToJsonFunction.java]: Please use only capital letters for function name. I guess thats the general convention we are following for other functions  though I dont know whether it has any functional significance or not.

File: phoenix-core/src/main/java/org/apache/phoenix/schema/json/PhoenixJson.java [hunk 2]
```diff
@@ -0,0 +1,355 @@
...
+            if (nodeTemp == null || nodeTemp.isMissingNode()) {
+                return null;
+            }
+            node = nodeTemp;
+        }
+        return new PhoenixJson(node, node.toString());
+    }
+
+    public int getJsonArrayLength() {
+        int count = 0;
```
Comment (AakashPradeep) [inline phoenix-core/src/main/java/org/apache/phoenix/schema/json/PhoenixJson.java]: Please add comment here, its hard to follow why we require this method and where we gonna use this. 

Probably it seems it will be used for json_array_length() function implementation. In that case first requirement will be to make sure that input is a valid Json array and if it is so then just check whether the rootnode (member variable of this class) isArray() or not, if it is an array then simply use size() method of ArrayNode class.

File: phoenix-core/src/main/java/org/apache/phoenix/schema/json/PhoenixJson.java [hunk 3]
```diff
@@ -0,0 +1,355 @@
...
+    }
+
+    public Object[] getJsonFields() {
+        List<String> elementlist = new ArrayList();
+        Iterator<Map.Entry<String, JsonNode>> fields = this.rootNode.getFields();
+
+        while(fields.hasNext()){
+            String fieldsstr = "";
+            Map.Entry<String, JsonNode> entry = fields.next();
+            fieldsstr += entry.getKey() +","+entry.getValue().toString();
```
Comment (AakashPradeep) [inline phoenix-core/src/main/java/org/apache/phoenix/schema/json/PhoenixJson.java]: Please specify why you are using "," as a separator b/w key and value.

File: phoenix-core/src/main/java/org/apache/phoenix/schema/json/PhoenixJson.java [hunk 4]
```diff
@@ -0,0 +1,355 @@
...
+        Iterator<String> fieldnames = this.rootNode.getFieldNames();
+        while(fieldnames.hasNext()){
+            elementlist.add(fieldnames.next());
+        }
+        return elementlist.toArray();
+    }
+
+    public Object[] getJsonFields() {
+        List<String> elementlist = new ArrayList();
+        Iterator<Map.Entry<String, JsonNode>> fields = this.rootNode.getFields();
```
Comment (AakashPradeep) [inline phoenix-core/src/main/java/org/apache/phoenix/schema/json/PhoenixJson.java]: Nice !!

File: phoenix-core/src/main/java/org/apache/phoenix/schema/json/PhoenixJson.java [hunk 5]
```diff
@@ -0,0 +1,355 @@
...
+
+        while(fields.hasNext()){
+            String fieldsstr = "";
```
Comment (AakashPradeep) [inline phoenix-core/src/main/java/org/apache/phoenix/schema/json/PhoenixJson.java]: Define variable only the places where it requires  it makes easier to read the code.

File: phoenix-core/src/main/java/org/apache/phoenix/schema/json/PhoenixJson.java (hunk 3 above)
Comment (AakashPradeep) [inline phoenix-core/src/main/java/org/apache/phoenix/schema/json/PhoenixJson.java]: If you define String variable here, you do not need to use "+" operator, which is a costly affair avoid using it as much as possible.

File: phoenix-core/src/main/java/org/apache/phoenix/schema/json/PhoenixJson.java [hunk 6]
```diff
@@ -0,0 +1,355 @@
...
+            Map.Entry<String, JsonNode> entry = fields.next();
+            fieldsstr += entry.getKey() +","+entry.getValue().toString();
+            elementlist.add(fieldsstr);
+        }
+        return elementlist.toArray();
+    }
+    public String PopulateRecord(String [] types) {
```
Comment (AakashPradeep) [inline phoenix-core/src/main/java/org/apache/phoenix/schema/json/PhoenixJson.java]: Its hard to understand why do you need this method and what it is suppose to do. So, please add comments and do not start a method name with a capital letter.

**Comment by ictwanglei** [comment]
@AakashPradeep  sorry for the missing of comments and other coding mistakes and thanks for your patient guidance. I have modified the code according to your suggestion.

**Comment by twdsilva** [comment]
@ictwanglei  Can you please rebase and upload new patch, this one does not apply cleanly to the json branch?

File: phoenix-core/src/it/java/org/apache/phoenix/end2end/JsonArrayLengthFunctionIT.java [hunk 7]
```diff
@@ -0,0 +1,114 @@
+/**
```
Comment (twdsilva) [inline phoenix-core/src/it/java/org/apache/phoenix/end2end/JsonArrayLengthFunctionIT.java]: Add the APACHE header

File: phoenix-core/src/it/java/org/apache/phoenix/end2end/JsonArrayLengthFunctionIT.java [hunk 8]
```diff
@@ -0,0 +1,114 @@
...
+					rs.getString(1));
+			assertFalse(rs.next());
+
+		} finally {
+			conn.close();
+		}
+	}
+
+	@Test
+	public void testJsonArrayLengthWhenFirstArgumentIsJsonString()
```
Comment (twdsilva) [inline phoenix-core/src/it/java/org/apache/phoenix/end2end/JsonArrayLengthFunctionIT.java]: Change test name to testJsonArrayWithDifferentDataTypes

File: phoenix-core/src/it/java/org/apache/phoenix/end2end/JsonArrayLengthFunctionIT.java [hunk 9]
```diff
@@ -0,0 +1,114 @@
...
+import org.junit.Test;
+
+/**
+ * End to end test for {@link JsonArrayLengthFunction}.
+ *
+ */
+public class JsonArrayLengthFunctionIT extends BaseHBaseManagedTimeIT {
+
+	@Test
+	public void testJsonArrayLengthWithWhereClause() throws Exception {
```
Comment (twdsilva) [inline phoenix-core/src/it/java/org/apache/phoenix/end2end/JsonArrayLengthFunctionIT.java]: Can you test with more complex json (like nested json and the different data types)

File: phoenix-core/src/it/java/org/apache/phoenix/end2end/JsonArrayLengthFunctionIT.java [hunk 10]
```diff
@@ -0,0 +1,114 @@
...
+
+			assertFalse(rs.next());
+
+		} finally {
+			conn.close();
+		}
+	}
+
+    @Test
+    public void testJsonArrayElementsWithWhereClause() throws Exception {
```
Comment (twdsilva) [inline phoenix-core/src/it/java/org/apache/phoenix/end2end/JsonArrayLengthFunctionIT.java]: Can you add a test that verifies json_array_elements returns an array correctly ?

File: phoenix-core/src/main/java/org/apache/phoenix/schema/json/PhoenixJson.java [hunk 11]
```diff
@@ -231,4 +236,197 @@ private PhoenixJson getPhoenixJsonInternal(String[] paths) {
...
+     * If the current {@link PhoenixJson} is a JsonArray,then it returns the length of the JsonArray.
+     * <p>For example:[1,2,3] ,it will return 3
+     * <p>Its required for json_array_length().
+     * @throws SQLException
+     */
+    public int getJsonArrayLength() throws SQLException {
+       if(this.rootNode.isArray()){
+           return this.rootNode.size();
+       }else{
+           throw new SQLException("The JsonNode should be an Array");
```
Comment (twdsilva) [inline phoenix-core/src/main/java/org/apache/phoenix/schema/json/PhoenixJson.java]: Add a new SQLExceptionCode and use SQLExceptionInfo.Builder

File: phoenix-core/src/main/java/org/apache/phoenix/schema/json/PhoenixJson.java [hunk 12]
```diff
@@ -231,4 +236,197 @@ private PhoenixJson getPhoenixJsonInternal(String[] paths) {
...
+        if(this.rootNode.isArray()) {
+            List<String> elementlist = new ArrayList();
+            Iterator<JsonNode> elements = this.rootNode.getElements();
+            while (elements.hasNext()) {
+                JsonNode e = elements.next();
+                elementlist.add(e.toString());
+            }
+            return elementlist.toArray();
+        }else{
+            throw new SQLException("The JsonNode should be an Array");
```
Comment (twdsilva) [inline phoenix-core/src/main/java/org/apache/phoenix/schema/json/PhoenixJson.java]: Add a new SQLExceptionCode and use SQLExceptionInfo.Builder

File: phoenix-core/src/main/java/org/apache/phoenix/expression/function/JsonArrayElementsFunction.java [hunk 13]
```diff
@@ -0,0 +1,107 @@
...
+import org.apache.phoenix.schema.tuple.Tuple;
+import org.apache.phoenix.schema.types.*;
+import org.apache.phoenix.util.ByteUtil;
+
+import java.sql.SQLException;
+import java.util.List;
+
+@BuiltInFunction(name = JsonArrayElementsFunction.NAME, args = {
+        @Argument(allowedTypes = { PJson.class })})
+public class JsonArrayElementsFunction extends ScalarFunction {
```
Comment (twdsilva) [inline phoenix-core/src/main/java/org/apache/phoenix/expression/function/JsonArrayElementsFunction.java]: The corresponding postgress function returns a SETOF json. Is there a way that can be supported?

File: phoenix-core/src/main/java/org/apache/phoenix/expression/function/ArrayToJsonFunction.java [hunk 14]
```diff
@@ -0,0 +1,105 @@
...
+import org.apache.phoenix.util.ByteUtil;
+
+import java.io.IOException;
+import java.sql.SQLException;
+import java.util.List;
+
+
+@FunctionParseNode.BuiltInFunction(name=ArrayToJsonFunction.NAME,  args={
+        @FunctionParseNode.Argument(allowedTypes={PVarchar.class})} )
+public class ArrayToJsonFunction extends ScalarFunction {
```
Comment (twdsilva) [inline phoenix-core/src/main/java/org/apache/phoenix/expression/function/ArrayToJsonFunction.java]: Please add tests for this function

File: phoenix-core/src/main/java/org/apache/phoenix/expression/function/JsonEachFunction.java [hunk 15]
```diff
@@ -0,0 +1,96 @@
...
+import org.apache.phoenix.schema.json.PhoenixJson;
+import org.apache.phoenix.schema.tuple.Tuple;
+import org.apache.phoenix.schema.types.*;
+
+import java.util.List;
+
+@BuiltInFunction(name = JsonEachFunction.NAME, args = {
+        @Argument(allowedTypes = { PJson.class })})
+public class JsonEachFunction extends ScalarFunction {
+    public static final String NAME = "JSON_EACH";
```
Comment (twdsilva) [inline phoenix-core/src/main/java/org/apache/phoenix/expression/function/JsonEachFunction.java]: Add some tests for this function.

File: phoenix-core/src/main/java/org/apache/phoenix/schema/json/PhoenixJson.java [hunk 16]
```diff
@@ -231,4 +236,197 @@ private PhoenixJson getPhoenixJsonInternal(String[] paths) {
...
+     * If a column is specified more than once, the last value is used.
+     * Also use "," to separate each columns
+     * <p>For example:types :{"a","b"} json: {"a":"1","b":"2"}
+     * it will return new String("1,2")
+     *
+     * <p>Its required for json_populate_record().
+     * @param types {@link String} the record type
+     * @return {@link String} as the result record
+     */
+    public String jsonPopulateRecord(String [] types) {
```
Comment (twdsilva) [inline phoenix-core/src/main/java/org/apache/phoenix/schema/json/PhoenixJson.java]: Maybe add a helper function that can be used by both jsonPopulateRecord and jsonPopulateRecordSet?

File: phoenix-core/src/test/java/org/apache/phoenix/expression/function/JsonFunctionTest.java [hunk 17]
```diff
@@ -0,0 +1,284 @@
...
+        ImmutableBytesWritable ptr = new ImmutableBytesWritable();
+        boolean evaluated = e.evaluate(null, ptr);
+        PhoenixJson result = (PhoenixJson)e.getDataType().toObject(ptr);
+        return result;
+    }
+
+    @Test
+    public void testNumberArrayToJson() throws Exception {
+        Object[] testarray = new Object[]{1,12,32,432};
+        PhoenixJson result = testArrayToJson(testarray, PInteger.INSTANCE, PIntegerArray.INSTANCE);
```
Comment (twdsilva) [inline phoenix-core/src/test/java/org/apache/phoenix/expression/function/JsonFunctionTest.java]: Just add test that executes this function as a query.

File: phoenix-core/src/test/java/org/apache/phoenix/expression/function/JsonFunctionTest.java [hunk 18]
```diff
@@ -0,0 +1,284 @@
...
+        testarray = new Object[]{t1,t2,t3};
+        result = testArrayToJson(testarray, PTime.INSTANCE,PTimeArray.INSTANCE);
+        expected ="[\"11:01:45\",\"13:01:45\",\"12:01:45\"]";
+        assertEquals(result.serializeToString(), expected);
+
+
+    }
+
+
+    private static  String[] jsonArrayElements (String json) throws Exception {
```
Comment (twdsilva) [inline phoenix-core/src/test/java/org/apache/phoenix/expression/function/JsonFunctionTest.java]: Add a test that executes this function as a query.

File: phoenix-core/src/test/java/org/apache/phoenix/expression/function/JsonFunctionTest.java [hunk 19]
```diff
@@ -0,0 +1,284 @@
...
+        List<Expression> children = Arrays.<Expression>asList(typesExpr,JsonExpr);
+        JsonPopulateRecordFunction e = new JsonPopulateRecordFunction(children);
+        ImmutableBytesWritable ptr = new ImmutableBytesWritable();
+        boolean evaluated = e.evaluate(null, ptr);
+        String record = (String)e.getDataType().toObject(ptr);
+        return record;
+    }
+
+    @Test
+    public void testJsonPopulateRecord() throws Exception {
```
Comment (twdsilva) [inline phoenix-core/src/test/java/org/apache/phoenix/expression/function/JsonFunctionTest.java]: Add a test that executes this function as a query.

File: phoenix-core/src/test/java/org/apache/phoenix/expression/function/JsonFunctionTest.java [hunk 20]
```diff
@@ -0,0 +1,284 @@
...
+        Object[] expected = new Object[]{"f2","f4","f7"};
+        String[] result = jsonObjectKeys(TEST_JSON_STR);
+
+        assertEquals(result.length, expected.length);
+        for(int i = 0; i<result.length;i++){
+            assertEquals(result[i], expected[i]);
+        }
+    }
+
+    private static String[] jsonEach (String json) throws Exception {
```
Comment (twdsilva) [inline phoenix-core/src/test/java/org/apache/phoenix/expression/function/JsonFunctionTest.java]: Add a test that executes this function as a query.
